20_classes_lab/03.py

```diff
@@ -1,13 +1,47 @@
-"""
-Following code assumes a class named Widget that depends on
-other widgets
-Widget ctor takes a name, and function 
-add_dependency(depends)
-marks widget as depends on all widgets in "depends" list
-
-When building a widget, you need to first build all the widgets it
-depends on.
-
-Implement widget so the following code works
-"""
-
+""" Class Widget: traverse dependency tree (recursion) """
+
+class Widget(object):
+
+    def __init__(self,name):
+        self.name = name
+        self.dependencies = []
+
+    def add_dependency(self,*widgets):
+        for widget in widgets: self.dependencies.append(widget)
+
+    def build_internal(self,name_list):
+        if self.name in name_list: return []
+        for dependency in self.dependencies:

```

Comment:
What about nested dependencies ? When A depends on B which depends on C and so on. A flat loop won't catch everything